Refactor event listeners

Chrizpy opened the pull request:
Related to: https://github.com/twingly/ecco/issues/25

Reduce duplication with the help of helper classes.

Chrizpy committed:
Add RowType

Related to the issue: https://github.com/twingly/ecco/issues/25
Try to come up with a helper class to reduce duplication

Chrizpy committed:
Implement RowType and RowTypeSave

Implement new helper classes and clean up code.

walro commented on lib/ecco/row_event_listener.rb:
```diff
@@ -1,35 +1,21 @@
 require "ecco/event_listener"
 require "ecco/row_event"
+require_relative "row_type"
```
this seems to break the pattern for requires, use:

```suggestion
require "ecco/row_type"
```

Chrizpy replied:
require pattern aligned: 02f8b78380454419d6fc90559eff9079c17e146d

walro replied:
Great, consider using "Commit suggestion" next time though :)

<img width="240" alt="Screenshot 2020-11-03 at 08 06 29" src="https://user-images.githubusercontent.com/2550510/97957443-80a76800-1dab-11eb-812c-4ef949736ca6.png">

Chrizpy replied:
Ahh sorry!

walro replied:
hehe this came back?

Chrizpy replied:
Huh, that is very odd! I'm not really sure how this happened, even when looking through the commits 😅
I got to try clicking the correct button though, so good practise nonetheless!

walro commented on lib/ecco/row_type.rb:
```diff
@@ -0,0 +1,42 @@
+      EventType::ROTATE => "ROTATE",
+    }
+
+    SAVE_EVENTS = SAVE_EVENTS.merge(EVENTS)
```
Not sure overwriting a constant is best practice, if you run with warnings won't it complain?

Chrizpy replied:
Hmm you're right it did. Should I build `SAVE_EVENTS` exactly like I do `EVENTS` then? I was doing it like this because I tried to limit even more how much duplication is done. It would be kind of neat if I could just include the other hash inside this hash here, but I don't think there is a standard way of doing that?

walro replied:
something like this ought to work:

```ruby
HASH = { some: "key" }.merge(SOME_OTHER_HASH)
```

Chrizpy replied:
Aha cool! Did that here now! 2b8667dae977854fb6e473ce499e014c25ece474

Chrizpy committed:
Fix require pattern

Chrizpy committed:
Properly initialize SAVE_EVENTS

walro commented on lib/ecco/row_type.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  class RowTypeSave < RowType
+    SAVE_EVENTS = {
```
Seeing it's only the constant that's different between this class and its parent, I have a hunch things could be simplified, I don't think one needs to implement `.accepted_events` and `.fetch_event` here again if refactored a bit

Chrizpy replied:
Hmm I think I know what to do.
EDIT: Nope, my idea involved tampering with an already initialized constant.

Chrizpy replied:
I'm investigating if this is a viable solution right now, it is currently not going through the spec though:

```
module Ecco
  class RowType
    java_import com.github.shyiko.mysql.binlog.event.EventType

    # MySQL v1 and v2 row events
    include_events = {
      EventType::WRITE_ROWS      => "WRITE_ROWS",
      EventType::EXT_WRITE_ROWS  => "WRITE_ROWS",
      EventType::UPDATE_ROWS     => "UPDATE_ROWS",
      EventType::EXT_UPDATE_ROWS => "UPDATE_ROWS",
      EventType::DELETE_ROWS     => "DELETE_ROWS",
      EventType::EXT_DELETE_ROWS => "DELETE_ROWS"
    }

    EVENTS = Hash.new.merge(include_events)

    TABLE_EVENT = EventType::TABLE_MAP

    def self.accepted_events
      EVENTS.keys
    end

    def self.fetch_event(type)
      EVENTS.fetch(type)
    end
  end

  class RowTypeSave < RowType
    include_events = {
      EventType::QUERY  => "QUERY",
      EventType::ROTATE => "ROTATE",
    }.merge(EVENTS)
  end
end
```

The problem I think is that when `RowTypeSave` inherits `RowType`, it does `EVENTS = Hash.new.merge(include_events)` straight away, and since `include_events` doesn't exist yet when this is done, the hash is instead merged with a null object.

walro replied:
I think you can override constans without warnings, this is OK:

```
class Parent
  CONSTANT = 1
end

class Child < Parent
  CONSTANT = 2
end

puts Parent::CONSTANT
=> 1

puts Child::CONSTANT
=> 2
```

Chrizpy replied:
Just tried this in pry, you're right! That makes things easier! :)

Chrizpy replied:
Now the child class only overrides the constant! 267f14e57d05f7f366f6b1b952aa0f013e984ad5

Chrizpy committed:
Further refactoring

Chrizpy committed:
Align the require pattern, again

Co-authored-by: Robin Wallin <walro467@gmail.com>

walro commented on lib/ecco/row_type.rb:
```diff
@@ -0,0 +1,32 @@
+module Ecco
+  class RowType
+    java_import com.github.shyiko.mysql.binlog.event.EventType
```
since this import was added here I think it can be removed in `event_listener.rb`:

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index a07bf4a..7efb318 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -1,7 +1,6 @@
 module Ecco
   class EventListener
     include com.github.shyiko.mysql.binlog.BinaryLogClient::EventListener
-    java_import com.github.shyiko.mysql.binlog.event.EventType
 
     attr_writer :callback
 ```

walro replied:
However, one can reason a bit whether some things that were moved to a completely new class (RowType) could be moved into the already existing parent class.

Chrizpy replied:
Ah, so skip RowType entirely and move in the two classes in RowType into EventListener?

walro replied:
Yeah, I think that could work out

Chrizpy replied:
So something like this? 67bfd8fcbaa3c53d731ab7527b6a4f5b5d59135b

walro replied:
Kind of, I don't think we need the inner class, also now we can work with instances instead of classes which I think could make things a bit more simpler even.

Chrizpy replied:
Are you suggesting that I could move `EVENTS` into an instance variable `@events` instead and then in `SaveEventListener` override it to fit the latter instead?

Chrizpy replied:
I can show what I mean here: 4b5f22cf3ac874f8642eea085e4c30be31db6d52

walro replied:
> Are you suggesting that I could move `EVENTS` into an instance variable `@events` instead and then in `SaveEventListener` override it to fit the latter instead?

Almost, I think it's more common to work with overloading of methods than variables though

Chrizpy replied:
Aha!

walro replied:
I can try to make a diff tomorrow to show how it could look. Getting late now :)

Chrizpy replied:
So something like this? e9551dfbab7cf682884ba33e53e32aac9d1c713a

Chrizpy replied:
> I can try to make a diff tomorrow to show how it could look. Getting late now :)

No worries! No need to look at the latest commit :D

walro replied:
I don't know if this is readable, but here's the full diff of an slightly alternative way of doing it. Basically I built on the ideas here, but I re-instated the old "arrays" (Constants) to make it more clear in the subclasses what kind of events they care about. This also has the upshot of not generating a new hash on every event. Also changed the name of the helper to something more fitting (IMHO):

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index a41c8d3..b1d5a12 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -3,36 +3,32 @@ module Ecco
     include com.github.shyiko.mysql.binlog.BinaryLogClient::EventListener
     java_import com.github.shyiko.mysql.binlog.event.EventType
 
+    WRITE_EVENTS  = { EventType::WRITE_ROWS => "WRITE_ROWS", EventType::EXT_WRITE_ROWS => "WRITE_ROWS" }
+    UPDATE_EVENTS = { EventType::UPDATE_ROWS => "UPDATE_ROWS", EventType::EXT_UPDATE_ROWS => "UPDATE_ROWS" }
+    DELETE_EVENTS = { EventType::DELETE_ROWS => "DELETE_ROWS" , EventType::EXT_DELETE_ROWS => "DELETE_ROWS" }
+    QUERY_EVENTS  = { EventType::QUERY => "QUERY"}
+    ROTATE_EVENTS = { EventType::ROTATE => "ROTATE"}
+
+    TABLE_EVENT = EventType::TABLE_MAP
+
     attr_writer :callback
 
-    def initialize(client)
+    def initialize(client, events_of_interest)
       @client = client
+      @events_of_interest = events_of_interest
       @callback = Proc.new {}
     end
 
-    TABLE_EVENT = EventType::TABLE_MAP
-
     def on_event(event)
       raise NotImplementedError
     end
 
-    def events
-      {
-        EventType::WRITE_ROWS      => "WRITE_ROWS",
-        EventType::EXT_WRITE_ROWS  => "WRITE_ROWS",
-        EventType::UPDATE_ROWS     => "UPDATE_ROWS",
-        EventType::EXT_UPDATE_ROWS => "UPDATE_ROWS",
-        EventType::DELETE_ROWS     => "DELETE_ROWS",
-        EventType::EXT_DELETE_ROWS => "DELETE_ROWS"
-      }
-    end
-
     def accepted_events
-      events.keys
+      @events_of_interest.keys
     end
 
-    def fetch_event(type)
-      events.fetch(type)
+    def event_type_to_string(type)
+      @events_of_interest.fetch(type)
     end
   end
 end
diff --git a/lib/ecco/row_event_listener.rb b/lib/ecco/row_event_listener.rb
index 5d78eb9..5dcc2fe 100644
--- a/lib/ecco/row_event_listener.rb
+++ b/lib/ecco/row_event_listener.rb
@@ -3,6 +3,12 @@ require "ecco/row_event"
 
 module Ecco
   class RowEventListener < EventListener
+    def initialize(client)
+      events_of_interest =
+        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS)
+      super(client, events_of_interest)
+    end
+
     def on_event(event)
       data = event.get_data
       type = event.get_header.get_event_type
@@ -14,7 +20,7 @@ module Ecco
         row_event          = Ecco::RowEvent.new
         row_event.table_id = data.get_table_id
         row_event.rows     = data.rows
-        row_event.type     = fetch_event(type)
+        row_event.type     = event_type_to_string(type)
 
         if @table_map_event
           table_event_data = @table_map_event.get_data
diff --git a/lib/ecco/save_event_listener.rb b/lib/ecco/save_event_listener.rb
index 476aa2f..41fd52c 100644
--- a/lib/ecco/save_event_listener.rb
+++ b/lib/ecco/save_event_listener.rb
@@ -2,22 +2,22 @@ require "ecco/event_listener"
 
 module Ecco
   class SaveEventListener < EventListener
+    def initialize(client)
+      events_of_interest =
+        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS).merge(QUERY_EVENTS).merge(ROTATE_EVENTS)
+      super(client, events_of_interest)
+    end
+
     def on_event(event)
       type = event.get_header.get_event_type
       case type
       when *accepted_events
         filename        = @client.get_binlog_filename
         position        = @client.get_binlog_position
-        event_type_name = fetch_event(type)
+        event_type_name = event_type_to_string(type)
 
         @callback.call(filename, position, event_type_name)
       end
     end
-    def events
-      {
-        EventType::QUERY  => "QUERY",
-        EventType::ROTATE => "ROTATE"
-      }.merge(super)
-    end
   end
 end
```

walro replied:
I am not super-happy about the merging going on, there are probably better ways to solve that. Also the value of `#keys` should probably be stashed for some improved performance

Chrizpy committed:
Move RowType into EventListener

Chrizpy committed:
Remove absolute path

RowType is now in the EventListener class so absolute path is no longer needed.

Chrizpy committed:
Remove row_type.rb

Chrizpy committed:
Remove subclass, make EVENTS instance variable

Use instance variables for the events instead, remove the subclasses "RowType" and "RowTypeSave"

Chrizpy committed:
EVENTS is methods rather than variables

walro commented:
I tweaked the PR title a bit to emphasize what we're doing rather than which files are touched.

Chrizpy committed:
PR feedback

Don't generate a new hash every event

Keep constants

Better name for helper method

Co-Authored-By: Robin Wallin <2550510+walro@users.noreply.github.com>

Chrizpy committed:
Add helper method to retrieve events

retrieve_save_events makes the long lines of merges less clunky and easier to scan.

walro commented on lib/ecco/save_event_listener.rb:
```diff
@@ -2,32 +2,31 @@
+    private
+
+    def retrieve_save_events
+      Hash.new.merge(WRITE_EVENTS)
+              .merge(UPDATE_EVENTS)
+              .merge(DELETE_EVENTS)
+              .merge(QUERY_EVENTS)
+              .merge(ROTATE_EVENTS)
```
Nice, but I think we should use the same helper method in both event handlers to keep it consistent. Also one could do something like this to avoid the list of merges (they are till there, but hidden):

```diff
index e59ccc1..6d72a17 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -15,7 +15,7 @@ module Ecco
 
     def initialize(client, events_of_interest)
       @client = client
-      @events_of_interest = events_of_interest
+      @events_of_interest = events_of_interest.inject(&:merge)
       @callback = Proc.new {}
     end
 
diff --git a/lib/ecco/row_event_listener.rb b/lib/ecco/row_event_listener.rb
index 5dcc2fe..6991c3a 100644
--- a/lib/ecco/row_event_listener.rb
+++ b/lib/ecco/row_event_listener.rb
@@ -4,8 +4,6 @@ require "ecco/row_event"
 module Ecco
   class RowEventListener < EventListener
     def initialize(client)
-      events_of_interest =
-        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS)
       super(client, events_of_interest)
     end
 
@@ -32,5 +30,11 @@ module Ecco
         @callback.call(row_event)
       end
     end
+
+    private
+
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS]
+    end
   end
 end
diff --git a/lib/ecco/save_event_listener.rb b/lib/ecco/save_event_listener.rb
index dca57a0..8281969 100644
--- a/lib/ecco/save_event_listener.rb
+++ b/lib/ecco/save_event_listener.rb
@@ -3,7 +3,6 @@ require "ecco/event_listener"
 module Ecco
   class SaveEventListener < EventListener
     def initialize(client)
-      events_of_interest = retrieve_save_events
       super(client, events_of_interest)
     end
 
@@ -21,12 +20,8 @@ module Ecco
 
     private
 
-    def retrieve_save_events
-      Hash.new.merge(WRITE_EVENTS)
-              .merge(UPDATE_EVENTS)
-              .merge(DELETE_EVENTS)
-              .merge(QUERY_EVENTS)
-              .merge(ROTATE_EVENTS)
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS, QUERY_EVENTS, ROTATE_EVENTS]
     end
   end
 end
```

Chrizpy replied:
Wow that's so cool! TIL!!

walro replied:
One could actually further simplify, by doing this:

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index e59ccc1..62ebd81 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -13,10 +13,10 @@ module Ecco
 
     attr_writer :callback
 
-    def initialize(client, events_of_interest)
-      @client = client
-      @events_of_interest = events_of_interest
-      @callback = Proc.new {}
+    def initialize(client)
+      @client     = client
+      @event_list = events_of_interest.inject(&:merge)
+      @callback   = Proc.new {}
     end
 
     def on_event(event)
@@ -24,11 +24,17 @@ module Ecco
     end
 
     def accepted_events
-      @events_of_interest.keys
+      @event_list.keys
     end
 
     def event_type_to_string(type)
-      @events_of_interest.fetch(type)
+      @event_list.fetch(type)
+    end
+
+    private
+
+    def events_of_interest
+      raise NotImplementedError
     end
   end
 end
diff --git a/lib/ecco/row_event_listener.rb b/lib/ecco/row_event_listener.rb
index 5dcc2fe..26731a8 100644
--- a/lib/ecco/row_event_listener.rb
+++ b/lib/ecco/row_event_listener.rb
@@ -3,12 +3,6 @@ require "ecco/row_event"
 
 module Ecco
   class RowEventListener < EventListener
-    def initialize(client)
-      events_of_interest =
-        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS)
-      super(client, events_of_interest)
-    end
-
     def on_event(event)
       data = event.get_data
       type = event.get_header.get_event_type
@@ -32,5 +26,11 @@ module Ecco
         @callback.call(row_event)
       end
     end
+
+    private
+
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS]
+    end
   end
 end
diff --git a/lib/ecco/save_event_listener.rb b/lib/ecco/save_event_listener.rb
index dca57a0..912709d 100644
--- a/lib/ecco/save_event_listener.rb
+++ b/lib/ecco/save_event_listener.rb
@@ -2,11 +2,6 @@ require "ecco/event_listener"
 
 module Ecco
   class SaveEventListener < EventListener
-    def initialize(client)
-      events_of_interest = retrieve_save_events
-      super(client, events_of_interest)
-    end
-
     def on_event(event)
       type = event.get_header.get_event_type
       case type
@@ -21,12 +16,8 @@ module Ecco
 
     private
 
-    def retrieve_save_events
-      Hash.new.merge(WRITE_EVENTS)
-              .merge(UPDATE_EVENTS)
-              .merge(DELETE_EVENTS)
-              .merge(QUERY_EVENTS)
-              .merge(ROTATE_EVENTS)
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS, QUERY_EVENTS, ROTATE_EVENTS]
     end
   end
 end
```

walro replied:
Also, I think we should take care of `#keys` not having to be evaluated every time, that could just be something like (in addition to diff above):


```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index 62ebd81..bd9e5d9 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -14,9 +14,10 @@ module Ecco
     attr_writer :callback
 
     def initialize(client)
-      @client     = client
-      @event_list = events_of_interest.inject(&:merge)
-      @callback   = Proc.new {}
+      @client          = client
+      @event_list      = events_of_interest.inject(&:merge)
+      @accepted_events = @event_list.keys
+      @callback        = Proc.new {}
     end
 
     def on_event(event)
@@ -24,7 +25,7 @@ module Ecco
     end
 
     def accepted_events
-      @event_list.keys
+      @accepted_events
     end
 
     def event_type_to_string(type)
```

Chrizpy replied:
Added all the ideas in the comments here: 3dba86e644920cc9c75b7458674f341393e9058c

Chrizpy committed:
Based on PR feedback

Simplify the 'mergening' instead of hiding it.
Store the keys of the events to increase performance.

Co-Authored-By: Robin Wallin <2550510+walro@users.noreply.github.com>

walro commented on lib/ecco/event_listener.rb:
```diff
@@ -3,15 +3,33 @@ class EventListener
-    def initialize(client)
+    def initialize(client, events_of_interest)
       @client = client
+      @event_list = events_of_interest.inject(&:merge)
```
Think I mentioned it IRL, but there's also `#reduce` which is the same as `#inject`. It's just a matter of taste of which one prefers the most, so if you like `#inject`, let's keep it.

Chrizpy replied:
I had forgotten about this, I agree that `#reduce` fits better in this case. :)

Force pushed that change 👍

walro commented:
I think you can (should) squash all commits into a single one.

walro commented on lib/ecco/event_listener.rb:
```diff
@@ -3,15 +3,33 @@ class EventListener
+      @event_list = events_of_interest.reduce(&:merge)
+      @accepted_events = @event_list.keys
       @callback = Proc.new {}
     end
```
One final suggestion, not sure if the new variable name is that much better than `event_list` though, but I like the grouping I did below at least :)

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index 235aba7..fc01849 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -15,8 +15,10 @@ module Ecco
 
     def initialize(client, events_of_interest)
       @client = client
-      @event_list = events_of_interest.reduce(&:merge)
-      @accepted_events = @event_list.keys
+
+      @events_of_interest = events_of_interest.reduce(&:merge)
+      @accepted_events    = @events_of_interest.keys
+
       @callback = Proc.new {}
     end
 
@@ -29,7 +31,7 @@ module Ecco
     end
 
     def event_type_to_string(type)
-      @event_list.fetch(type)
+      @events_of_interest.fetch(type)
     end
   end
 end
```

Chrizpy replied:
Nice! ce5be7ea4e3cd88efcde17f9e6b5647a43a19682

Chrizpy committed:
PR Feedback

Grouping and changing back to old variable name

Co-Authored-By: Robin Wallin <2550510+walro@users.noreply.github.com>

walro reviewed (approved):
LGTM! Remember to squash it all together!
